### DISCUSSION TITLE
testing with flask overwrites global Registry?

### DISCUSSION THREAD

**Discussion by gjcarneiro** [hynek/svcs#84] [Q&A] [answered]
I have  a global svcs.Registry object for my flask App.

In tests, I use `svcs.flask.overwrite_value()` to mock a service.  It works.  

However, after the test runs, my mock service is still registered.  This is a big problem.  `svcs.flask.overwrite_*` doesn't seem to clean up after itself.

Looking at the code, I cannot see how this could possibly work, since `svcs.flask.overwrite_*`  changes the global Registry.

This might be a bug, I'm not sure.  I expected `svcs.flask.overwrite_*` to overwrite a service in the current Container only, not in the global Registry.

**Comment by hynek** [comment]
Have you tried registering the mock in the container-local registry? Overwrite predates this feature.

**Reply by gjcarneiro** [reply]
Hm, it sounds like this is exactly what I need!  I assume you mean `Container.register_local_value()`.  I wasn't aware of it.

Shouldn't `svcs.flask.overwrite_value()` use this?   It feels like `Container.register_local_value()` is a lower level API and that probably `svcs.flask.overwrite_value()` should be modified instead?

**Reply by hynek** [reply, marked as answer]
Container-local registries predate `overwrite_value` and they have different semantics so using it would be a breaking change.

I think the more correct way would be to deprecate overwrite_value and maybe in Flask's case (since it has global side-effects) add a new convenience API. But really, it's just `svcs.flask.container.register_local_XXX()`, no?

I'm not super-keen to add new APIs before settling #70, though.

**Reply by gjcarneiro** [reply]
Yes, I'm reasonably happy with `svcs.flask.container.register_local_XXX()`.  I just wish that the documentation recommended this instead of `svcs.flas.overwrite_XXX()` for testing.

**Reply by hynek** [reply]
Yes, but: what if you already registered a container-local service and want to overwrite that?

But I agree that they should be documented as the primary way to achieve this.
